### ISSUE TITLE
Installation issue: boost@1.70.0 on Fedora 35

### ISSUE THREAD

**Issue by pbrady** [LLNL/spack#28273]
### Steps to reproduce the issue

On an installation of fedora 35:
```shell
$ spack install boost@1.70.0 cxxstd=17 %gcc9.4.0
```
The outcome is a variety of error messages like:

```shell
 >> 16215    ./boost/thread/pthread/thread_data.hpp:60:5: error: missing binary operator before token "("
     16216       60 | #if PTHREAD_STACK_MIN > 0
...
```

### Information on your system

```shell
$ spack debug report
* **Spack:** 0.17.1-730-a4f0fbafbb
* **Python:** 3.10.1
* **Platform:** linux-fedora35-broadwell
* **Concretizer:** clingo
```

### Additional information

[spack-build-out.txt](https://github.com/spack/spack/files/7818378/spack-build-out.txt)
[spack-build-env.txt](https://github.com/spack/spack/files/7818379/spack-build-env.txt)

@hainest 

As far as I can tell this is an issue related to recent changes in glibc.  Here's how they fixed it in other package managers: https://github.com/Ma27/nixpkgs/commit/87a628b1027363e8756c8bef7f6e4b71f190ee54

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

**Comment by hainest** [comment]
@pbrady As much as I'd like to avoid adding yet another patch, it looks like we'll need to use the one you linked to. Have you tried building with that patch?

**Comment by pbrady** [comment]
I haven't tried yet but I can give it a whirl tomorrow and put together a pull request if it works

**Comment by pbrady** [comment]
@hainest I have a problem I'm not sure how to solve.  It seems that compilers I install with spack continue to use my system headers.  I say this because when I was on fedora 34, I could build `boost@1.70.0 %gcc@9.4.0` without any problems.  But on f35, the command triggers the above errors. So what is the correct way to restrict the patch?

If I do ` patch("pthread-stack-min-fix.patch", when="@1.69.0:1.72.0%gcc@11:")` then it compiles fine with my system compiler but still fails if I build via gcc@9.4.0 since it's pulling in my system headers.  I can loosen the restriction to something `%gcc@8:` but then it will be needlessly applying the patch to many non-broken versions.

What's the best path here?

**Comment by pbrady** [comment]
Actually, restricting on the compiler doesn't work since clang uses the the same system pthread headers (or has the same version) so on fedora35, the following also fails and should be patched:

`spack install boost@1.70.0 cxxstd=17 %clang`

**Comment by hainest** [comment]
I'm doing some archaeology. I'll get back to you later today.

**Comment by pbrady** [comment]
Thanks @hainest.  I've posted some questions on slack as well

**Comment by hainest** [comment]
This is ultimately a glibc issue, so it can affect every version of boost.thread and any compiler. Because spack doesn't (yet) model glibc as a dependency, we'll need to broadly apply this patch to affected versions.

The numerical check for PTHREAD_STACK_MIN was introduced by https://github.com/boostorg/thread/commit/e54f7e39 on 2018-08-31. This would first be released in version 1.69.0. That change was reverted by https://github.com/boostorg/thread/commit/74fb0a26 on 2019-11-20, corresponding to version 1.72.0.

These versions are the same as those covered by the nixpkgs fix. I'd say let's do the same.

`patch("pthread-stack-min-fix.patch", when="@1.69.0:1.73.0")` should do.
